Generator of tensor inputs with variable layout and structure (batch/non-batch, hybrid/non-hybrid, block/non-block)

pearu opened the pull request:
This PR introduces `TestCase.generate_simple_inputs` method that is an improved and generalized version of the `TestSparseCompressed._generate_small_inputs` method.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89502
* #88733
* __->__ #88914



cc @nikitaved @cpuhrsch @amjames @bhosmer

pearu committed:
Generator of tensor inputs with variable layout and structure (batch/non-batch, hybrid/non-hybrid, block/non-block)

[ghstack-poisoned]

pearu committed:
Update on "Generator of tensor inputs with variable layout and structure (batch/non-batch, hybrid/non-hybrid, block/non-block)"




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu committed:
Update on "Generator of tensor inputs with variable layout and structure (batch/non-batch, hybrid/non-hybrid, block/non-block)"




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu committed:
Update on "Generator of tensor inputs with variable layout and structure (batch/non-batch, hybrid/non-hybrid, block/non-block)"




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu committed:
Update on "Generator of tensor inputs with variable layout and structure (batch/non-batch, hybrid/non-hybrid, block/non-block)"




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu committed:
Update on "Generator of tensor inputs with variable layout and structure (batch/non-batch, hybrid/non-hybrid, block/non-block)"




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu committed:
Update on "Generator of tensor inputs with variable layout and structure (batch/non-batch, hybrid/non-hybrid, block/non-block)"




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu committed:
Update on "Generator of tensor inputs with variable layout and structure (batch/non-batch, hybrid/non-hybrid, block/non-block)"




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2439,6 +2440,265 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
+                    size = kwargs.pop('size', args[0].shape)  # to ensure that a zero-sized tensor has the desired shape
+                    assert len(args) == 1
+                    yield args[0].reshape(size)
+                elif layout is torch.sparse_coo:
+                    yield torch.sparse_coo_tensor(*args, **kwargs).coalesce()
```
Testing uncoalesced tensors is of value.

pearu replied:
`coalesce()` method call is now removed.

nikitaved commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2439,6 +2440,265 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
+            if isinstance(base, list):
+                return [expand_values(b, densesize) for b in base]
+
+            r = 0
+            for s in reversed(densesize):
+                r = [copy.deepcopy(r) for _ in range(s)]
```
Hmm, why can't we just create `r` as `torch.zeros(densesize)`?

nikitaved replied:
It does not work with tensors directly? Also, the Tensor class has a `tolist` method that preserves it's nesteness.

nikitaved replied:
I am not sure about that, now after looking at the test... And I am not sure we ever construct sparse tensors from something which is not TensorLike in ATen...

pearu replied:
You can construct tensors from lists, for instance:
```python
>>> torch.sparse_csr_tensor([0, 2, 4], [0, 1, 0, 1], [1, 2, 3, 4])
tensor(crow_indices=tensor([0, 2, 4]),
       col_indices=tensor([0, 1, 0, 1]),
       values=tensor([1, 2, 3, 4]), size=(2, 2), nnz=4,
       layout=torch.sparse_csr)
```
In tests and examples, we often construct tensors from Python lists, so I don't see the relevance of constructing these within ATen.

nikitaved replied:
If using tensors leads to cleaner code, and when there is `tolist`, then why not?

pearu replied:
> What about we create r as torch.zeros(densesize)? It appears that the output of this function is fed into torch.tensor anyway...

Done.

nikitaved commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2439,6 +2440,265 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
+                    assert 0  # unreachable
+            return
+
+        def expand_values(base, densesize):
+            """
+            Expand values in base to blocks of values with densesize.
+            """
```
I am confused. `base` seems to be a scalar.

nikitaved commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2439,6 +2440,265 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
             x = x.detach().clone()._coalesced_(False)
         return x, x._indices().clone(), x._values().clone()
 
+    def generate_simple_inputs(self, layout,
+                               device=None,
+                               dtype=None,
+                               index_dtype=None,
+                               enable_batched=True,
+                               enable_hybrid=True,
+                               enable_zero_sized=True,
+                               enable_batched_variable_nse=False,
+                               output_tensor=True,
+                               patterns=None):
```
What do you think about having `low/high` parameters for values?

pearu replied:
The aim is to have inputs where all values are distinct (to ease debugging and the development of sparse algorithms). I think `low/high` parameters make sense for randomly generated samples but here we want the inputs to be highly predictable.

nikitaved replied:
Knowing the range of values could be useful to control precision. The values generated seem to be only unique when each size `<10`, which is the case I assume, right?

pearu replied:
yes, the generated values aim at being unique. Notice the sizes of the generated tensors are intentionally small so that a human could easily track how the tensor elements transform when applying sparse tensor conversion/indexing and other tools that act mostly on the tensor structure rather than on the values.

For tests where precision parameters matter, one should use the `make_tensor` or other tensor generators that provide the low/high parameter feature.

nikitaved commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2439,6 +2440,265 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
+                r = [copy.deepcopy(r) for _ in range(s)]
+            if base == 0:
+                return r
+            for index in itertools.product(*map(range, densesize)):
+                v = base + sum(10**c * i for c, i in enumerate(index))
+                r1 = r
+                for i in range(len(densesize) - 1):
+                    r1 = r1[index[i]]
+                assert r1[index[-1]] == 0, r1[index[-1]]
+                r1[index[-1]] = v
+            return r
```
Maybe something like that is a bit easier to understand? Along with the comment about what kind of mapping is employed...
```python
In [3]: l = expand_values(1, (3, 4, 5))
                                                    
In [4]: torch.tensor(l)                             
Out[4]:                                
tensor([[[  1, 101, 201, 301, 401],                 
         [ 11, 111, 211, 311, 411],   
         [ 21, 121, 221, 321, 421],                                                                      
         [ 31, 131, 231, 331, 431]],                
                                                    
        [[  2, 102, 202, 302, 402],                 
         [ 12, 112, 212, 312, 412],          
         [ 22, 122, 222, 322, 422],                 
         [ 32, 132, 232, 332, 432]],         
                                                    
        [[  3, 103, 203, 303, 403],                                                                                                     
         [ 13, 113, 213, 313, 413], 
         [ 23, 123, 223, 323, 423],                 
         [ 33, 133, 233, 333, 433]]])
                                                    
In [5]: densesize = (3, 4, 5)       
                                                    
In [6]: x = torch.tensor(range(densesize[0]))       
                                                    
In [7]: for i, d in enumerate(densesize[1:]):       
   ...:     y = torch.tensor(range(d))       
   ...:     x = x[..., None] + (10 ** (i + 1)) * y[None, ...]                                                                                                 
   ...:                                      
                                                    
In [8]: x + 1                                                                                            
Out[8]:                                                             
tensor([[[  1, 101, 201, 301, 401],                                 
         [ 11, 111, 211, 311, 411],                                 
         [ 21, 121, 221, 321, 421],                                 
         [ 31, 131, 231, 331, 431]],                                           
                                                                    
        [[  2, 102, 202, 302, 402],                                            
         [ 12, 112, 212, 312, 412],                                            
         [ 22, 122, 222, 322, 422],                                            
         [ 32, 132, 232, 332, 432]],                                           

        [[  3, 103, 203, 303, 403],                                            
         [ 13, 113, 213, 313, 413],                                            
         [ 23, 123, 223, 323, 423],                                            
         [ 33, 133, 233, 333, 433]]])              
```

pearu replied:
Now using:
```python
        def expand_values(base, densesize):
            """
            Expand values in base to blocks of values with densesize.
            """
            if not densesize:
                return base
            if isinstance(base, list):
                return torch.stack([expand_values(b, densesize) for b in base])
            if base == 0:
                return torch.zeros(densesize, dtype=torch.int64)
            r = torch.tensor(range(densesize[0]))
            for i, d in enumerate(densesize[1:]):
                y = torch.arange(d, dtype=torch.int64) * (10 ** (i + 1))
                r = r[..., None] + y[None, ...]
            r.add_(base)
            return r
```

pearu committed:
Update on "Generator of tensor inputs with variable layout and structure (batch/non-batch, hybrid/non-hybrid, block/non-block)"




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu committed:
Update on "Generator of tensor inputs with variable layout and structure (batch/non-batch, hybrid/non-hybrid, block/non-block)"




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2448,6 +2448,273 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
+                                target[j].append(d[j])
+            return batch_data
+
+        def expand_values(base, densesize):
+            """
+            Expand values in base to blocks of values with densesize.
+            """
```
```suggestion
        def generate_values(offset, densesize):
            """
            Generates a tensor of shape densesize with values
            at indices i_1, ..., i_d with i_j < densesize[j] for any 0 < j < len(densesize)
            equal to i_1 * 10^0 + ... + i_d * 10^{d - 1}.
            This mapping produces UNIQUE values as long as densesize[i] < 10.
            """
```
?

pearu replied:
Applied with some modifications.

nikitaved commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2448,6 +2448,273 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
+            return ()
+
+        def get_blockpattern(pattern, blocksize):
+            basesize = get_shape(pattern)
```
nit: if pattern is a tensor, then there is no need in `get_shape`.

pearu replied:
generate_simple_inputs now uses tensors

nikitaved commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2448,6 +2448,273 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
+            assert basesize[0] % blocksize[0] == 0, (basesize, blocksize)
+            assert basesize[1] % blocksize[1] == 0, (basesize, blocksize)
+            blockpattern = [[0 for _ in range(basesize[1] // blocksize[1])] for _ in range(basesize[0] // blocksize[0])]
+            for i, row in enumerate(pattern):
+                for j, maskin in enumerate(row):
+                    bi = i // blocksize[0]
+                    bj = j // blocksize[1]
+                    blockno = 1 + bi * (basesize[1] // blocksize[1]) + bj
+                    blockpattern[bi][bj] = blockno if maskin else blockpattern[bi][bj]
```
There could be several `i, j` that are mapped to the same `bi, bj`, which means that only the very last "valid" value will be written. It is not a concern here, right? Also nit: seems like this could be simplified with tensors and broadcasting.

pearu replied:
generate_simple_inputs now uses tensors

nikitaved commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2448,6 +2448,273 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
+            crow_indices = [0 for _ in range(basesize[0] + 1)]
+            ccol_indices = [0 for _ in range(basesize[1] + 1)]
+            strided_values = [[0 for _ in range(basesize[1])] for _ in range(basesize[0])]
+            for i, row in enumerate(pattern):
+                for j, maskin in enumerate(row):
+                    if maskin:
+                        count += 1
+                        indices.append([i, j])
+                        values.append(count)
+                        strided_values[i][j] = count
+                        for i1 in range(i, basesize[0]):
+                            crow_indices[i1 + 1] += 1
+                        for j1 in range(j, basesize[1]):
+                            ccol_indices[j1 + 1] += 1
```
nit: We can probably simplify this with tensors + masking + advanced indexing.

pearu replied:
generate_simple_inputs now uses tensors

pearu committed:
Update on "Generator of tensor inputs with variable layout and structure (batch/non-batch, hybrid/non-hybrid, block/non-block)"




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

amjames commented on test/test_sparse.py:
```diff
@@ -4049,6 +4049,58 @@ def test_basic(self):
+
+    def test_generate_simple_inputs(self):
+        # Temporarily disable BSC and BSC layouts as these don't support select yet
+        layouts = [torch.strided, torch.sparse_coo, torch.sparse_csr, torch.sparse_csc, torch.sparse_bsr, torch.sparse_bsc][:-2]
```
Is the select requirement coming from `assertEqual`? 

Is there some workaround we could use to test the generator for those layouts? I don't feel great about them being available to pull from the generator without testing.

amjames commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2448,6 +2448,285 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
+            #   patterns of the generated inputs: zero values
+            #   correspond to unspecified elements/blocks and non-zero
+            #   values the specified elements while elements with the
+            #   same value belong to the same block in some cases.
```
> `same value belong to the same block in some cases.`

This is a bit confusing what are the conditions for "some cases" and what happens in the other cases (where specified values do not fall in the same block, or the layout is not blocked)?

pearu replied:
I have updated the comments.

amjames reviewed (commented):
One minor comment inline. Since this is pretty complicated I think it is important that the comments are very clear.

pearu committed:
Update on "Generator of tensor inputs with variable layout and structure (batch/non-batch, hybrid/non-hybrid, block/non-block)"

This PR introduces `TestCase.generate_simple_inputs` method that is an improved and generalized version of the `TestSparseCompressed._generate_small_inputs` method.




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

cpuhrsch commented on test/test_sparse.py:
```diff
@@ -4049,6 +4049,58 @@ def test_basic(self):
+        tested_combinations = set()
+        while True:
+            try:
+                tensors = [next(g) for g in generators]
```
Can't you use [`zip`](https://docs.python.org/3.3/library/functions.html#zip) for this? Or is that not the intend?

pearu replied:
Yes, using `zip` is a good idea.

pearu replied:
Done. Thanks for the suggestion, @cpuhrsch !

pearu committed:
Update on "Generator of tensor inputs with variable layout and structure (batch/non-batch, hybrid/non-hybrid, block/non-block)"

This PR introduces `TestCase.generate_simple_inputs` method that is an improved and generalized version of the `TestSparseCompressed._generate_small_inputs` method.




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu committed:
Update on "Generator of tensor inputs with variable layout and structure (batch/non-batch, hybrid/non-hybrid, block/non-block)"

This PR introduces `TestCase.generate_simple_inputs` method that is an improved and generalized version of the `TestSparseCompressed._generate_small_inputs` method.




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

cpuhrsch commented:
@pytorchbot merge
